[AIRFLOW-5911] Simplify lineage API and improve robustness (#6564)
mistercrunch/Airflow: docs/lineage.rst

    f_final = File("/tmp/final")
    run_this_last = DummyOperator(task_id='run_this_last', dag=dag,
        inlets=AUTO,
        outlets=f_final)

    f_in = File("/tmp/whole_directory/")
    outlets = []
    for file in FILE_CATEGORIES:
        f_out = File("/tmp/{}/{{{{ execution_date }}}}".format(file))
        outlets.append(f_out)

    run_this = BashOperator(
        task_id='run_me_first', bash_command='echo 1', dag=dag,
        inlets=f_in,
        outlets=outlets
    )
    run_this.set_downstream(run_this_last)

Tasks take the parameters ``inlets`` and ``outlets``.

Inlets can be manually defined by the following options:

> - by a list of dataset

Reviewer: ```suggestion
- by a list of airflow.lineage.entities.Dataset or a subclass
```

> - by a list of airflow.lineage.entities.Dataset or a subclass

Reviewer 2: ```suggestion
- by a list of ``airflow.lineage.entities.Dataset`` or a subclass
```

- can be configured to look for outlets from upstream tasks

> - can be configured to pick up outlets from direct upstream tasks 'AUTO'

Reviewer: The way this is written people would do `inlets="AUTO"` which won't be right.

```suggestion
- can be configured to pick up outlets from direct upstream tasks ``airflow.lineage.AUTO``
```

Author: Should be 'auto' & 'AUTO' & 'airflow.lineage.AUTO'.

- a combination of them

> Outlets are defined as list of dataset ``[dataset1, dataset2]}``.

Reviewer: ```suggestion
Outlets are defined as list of datasets ``[dataset1, dataset2]``.
```

All fields for datasets are templated with the context when the task is being executed. They will be templated during
execution of the originating task and willnot be re-templated if picked up downstream.

.. note:: Operators can add inlets and outlets automatically if the operator supports it.

> In the example DAG task ``run_this``(task_id ``run_me_first``) is a BashOperator that takes 3 inlets: ``CAT1``, ``CAT2``, ``CAT3``, that are

Reviewer 2: ```suggestion
In the example DAG task ``run_this``(task_id=``run_me_first``) is a BashOperator that takes 3 inlets: ``CAT1``, ``CAT2``, ``CAT3``, that are
```
